network/server.go

```diff
@@ -851,3 +892,47 @@ func (s *Server) persistLastKnownPeers() {
 		s.logger.Warn("Failed to persist lastKnownPeers to file", "err", err)
 	}
 }
+
+func (s *Server) monitorBootnodeFailures() {
+	ctx := context.Background()
+	eventCh, err := s.SubscribeCh(ctx)
+	if err != nil {
+		s.logger.Error("Failed to subscribe to peer events for bootnode fallback", "err", err)
+		return
+	}
+	for ev := range eventCh {
+		if ev.Type == peerEvent.PeerFailedToConnect {
+			s.bootnodeFallbackLock.Lock()
+			_, isBootnode := s.bootnodes.bootnodesMap[ev.PeerID]
+			if isBootnode {
+				s.failedBootnodeIDs[ev.PeerID] = struct{}{}
+				if len(s.failedBootnodeIDs) == len(s.bootnodes.bootnodeArr) && !s.bootnodeFallbackUsed && len(s.bootnodes.bootnodeArr) > 0 {
+					s.logger.Warn("All bootnode connection attempts failed, falling back to last_peers.json")
+					s.bootnodeFallbackUsed = true
+					// Clear current bootnodes
+					s.bootnodes.bootnodeArr = nil
```

Comment:
It will be better to keep the main bootnodes in the list as many of them will return online at some point